### PR TITLE
reportAgg now allows combined dimension mapping.

### PR DESCRIPTION
**Pull request by robinhasse**
- You can now define multi-dimensional sets in the BRICK reporting template.
- `readBrickSets` will auto-fill elements of those sets from the combination of the primary sets if elements are not provided explicitly.
- If reporting names are missing, they are also built from the primary reporting names separated by `|`. If a multi-dimensional set entry consists only of identical set elements, the reporting name won't be repeated but used just once. 
  - This helps e.g. to report identical replacement.
- `reportAgg` now takes multi-dimensional sets in `agg` and `rprt` separated by `.` in the following form: `agg = c(bs = "all", hs.hsr = "biom.gabo", ...)`

### CODE REVIEW

File: R/reportAgg.R [hunk 1]
```diff
@@ -282,3 +301,56 @@ reportAgg <- function(x,
...
+      names(patternRaw) <- tail(getSets(data), -2)
+      regex <- paste(unlist(lapply(dims[[cd]], function(elem) {
+        pattern <- patternRaw
+        pattern[.split(cd)] <- str_escape(.split(elem))
+        paste(pattern, collapse = "\\.")
+      })), collapse = "|")
+      items <- grep(regex, getItems(x, 3), value = TRUE)
+      data[, , items]
+    })
+  }
```
Comment (ricardarosemann) [inline R/reportAgg.R:353]: As discussed, these lines might also profit from additional comments, e.g. explaining the purpose of ```patternRaw```.

File: R/convGDX2MIF.R [hunk 2]
```diff
@@ -96,9 +96,12 @@ convGDX2MIF <- function(gdx,
 
 .findInconsistenSetElements <- function(brickSets, gdx) {
   m <- Container$new(gdx)
-  setsGdx <- setNames(m$getSymbols(names(brickSets)), names(brickSets))
+  sets <- unique(.split(names(brickSets)))
```
Comment (ricardarosemann) [inline R/convGDX2MIF.R:99]: As for the comment on ```.split``` in ```reportBrickSet.R```: It is hard to find the definition of ```.split``` (as well as ```.combinations``` below) as they are not found in the file (which is what I would usually expect). But as mentioned below, I'm also unsure what's the best way to handle this.

**Review by ricardarosemann — CHANGES_REQUESTED** [review]
Thanks a lot for implementing this, will be very useful.

I only have some minor remarks. Due to the complexity of the code I think it would be very helpful to add some comments to make it easier to understand; I marked several instances where this applies.

Furthermore, I'm a bit unhappy about the helper functions ```.split``` and ```.combinations``` being called in different files;  I also marked this at the respective occurrences. I don't really have a solution for this, though.

File: R/readBrickSets.R [hunk 3]
```diff
@@ -50,6 +58,37 @@ readBrickSets <- function(tmpl) {
                                                 childDims))
   }
 
+  # auto-fill combined dimensions
+  combinedDims <- grep("\\.", names(brickSetsExplicit), value = TRUE)
+  for (cd in combinedDims) {
+    dims <- .split(cd)
```
[the remark this answers is not in the document]
Comment (robinhasse) [reply R/readBrickSets.R:84]: Thanks, I agree. I have moved `.split` to the new script 'helper.R' where we can collect all small functions not worth giving an individual file that are used in more than one bigger function.

File: R/readBrickSets.R [hunk 4]
```diff
@@ -50,6 +58,37 @@ readBrickSets <- function(tmpl) {
...
+    if (identical(brickSetsExplicit[[cd]][["elements"]], NULL)) {
```
[the remark this answers is not in the document]
Comment (robinhasse) [reply R/readBrickSets.R]: I though there was but you are right, there isn't. `is.null` has always length 1. So I used it.

File: R/readBrickSets.R [hunk 5]
```diff
@@ -50,6 +58,35 @@ readBrickSets <- function(tmpl) {
...
+      # add all combinatinos of primary dimensions as elements
+      brickSetsExplicit[[cd]][["elements"]] <-
+        .combinations(lapply(dims, function(dim) {
+          names(brickSetsExplicit[[dim]][["elements"]])
+        }))
+    }
+    if (!is.list(brickSetsExplicit[[cd]][["elements"]])) {
```
[the remark this answers is not in the document]
Comment (robinhasse) [reply R/readBrickSets.R:92]: I want to allow defining elements just as an unnamed vector of elements without reporting names. In fact, the first step of the auto fill (the lines above) of entirely missing elements (the lines above) also does just this. Only if a list is provided, we assume that both elements and the respective reporting names are given and the latter don't need to be filled.

File: R/reportAgg.R [hunk 6]
```diff
@@ -87,10 +94,16 @@ reportAgg <- function(x,
   } else {
     # combination of entries of reporting dimensions
     rprtCombinations <- do.call(expand.grid, map$rprt)
+    for (cd in grep("\\.", colnames(rprtCombinations), value = TRUE)) {
+      if (cd %in% tagsInName) {
+        next
+      }
+      rprtCombinations <- rprtCombinations %>%
+        separate(cd, .split(cd))
+    }
```
[the remark this answers is not in the document]
Comment (robinhasse) [reply R/reportAgg.R]: I added some comments and hope that makes it clear now. I also saw that `separate` is superseded but decided to still go with it as long as the alternative remains experimental.

File: R/reportAgg.R (hunk 1 above)
Comment (robinhasse) [reply R/reportAgg.R]: I added in-line documentation and hope this makes it clearer.

File: R/convGDX2MIF.R (hunk 2 above)
Comment (robinhasse) [reply R/convGDX2MIF.R:99]: moved to helper.R

File: R/reportAgg.R [hunk 7]
```diff
@@ -227,16 +242,23 @@ reportAgg <- function(x,
...
+    if (!all(.split(dim) %in% getSets(x))) {
+      stop("x has no dimension(s) called ", paste(.split(dim), collapse = ", "))
+    }
+    if (grepl("\\.", dim)) {
+      existingElements <- unique(sub(
+        pattern = paste(rep("(.*)", magclass::ndim(x, 3)), collapse = "\\."),
+        replacement = paste(paste0("\\", match(.split(dim), tail(getSets(x), -2))), collapse = "\\."),
+        x = getItems(x, dim = 3)
+      ))
+      setdiff(dimLst[[dim]], existingElements)
```
[the remark this answers is not in the document]
Comment (robinhasse) [reply R/reportAgg.R]: Good catch. I accidently removed the original case of missing primary set elements. Should be fine now.

**Review by ricardarosemann — APPROVED** [review]
Thanks for addressing all remaining issues, looks good now :-)

**Pull request merged**